### PR TITLE
Add worker_key parameter to Adaptive

### PR DESCRIPTION
**Pull request by mrocklin**
This allows adaptive clusters to intelligently close down groups of
workers based on some logical association.

See https://github.com/dask/dask-jobqueue/pull/63 for motivation

cc @jhamman

### CODE REVIEW

**Comment by mrocklin** [comment]
@jhamman any luck with this?  My current plan is to wait on you to try things.  No rush, I just wanted to make sure that you weren't waiting on me to merge this.

**Comment by jhamman** [comment]
@mrocklin - So far so good. I'm testing this branch along with my mods in dask/dask-jobqueue#63 and things are working much better. I'll give this PR a quick review now.

File: distributed/deploy/adaptive.py [hunk 1]
```diff
@@ -223,7 +227,8 @@ def workers_to_close(self, **kwargs):
     @gen.coroutine
     def _retire_workers(self, workers=None):
         if workers is None:
-            workers = self.workers_to_close()
+            workers = self.workers_to_close(key=self.worker_key,
+                                            minimum=self.minimum)
```
Comment (jhamman) [inline distributed/deploy/adaptive.py]: what is the logic for passing in minimum here? Should you also pass in maximum?

File: distributed/deploy/local.py [hunk 2]
```diff
@@ -145,12 +146,24 @@ def __repr__(self):
...
+        asynchronous = kwargs.pop('asynchronous', None)
+        if asynchronous or self._asynchronous or getattr(thread_state, 'asynchronous', False):
+            callback_timeout = kwargs.pop('callback_timeout', None)
+            future = func(*args, **kwargs)
+            if callback_timeout is not None:
+                future = gen.with_timeout(timedelta(seconds=callback_timeout),
+                                          future)
+            return future
+        else:
+            return sync(self.loop, func, *args, **kwargs)
```
Comment (jhamman) [inline distributed/deploy/local.py]: Is this related to the grouped worker issue? (I'm not following)

**Comment by jhamman** [comment]
@mrocklin - save my few questions above, this seems to working and I'd like to see it merged in the next few days. Currently, this is blocking https://github.com/dask/dask-jobqueue/pull/63.

File: distributed/deploy/adaptive.py (hunk 1 above)
Comment (mrocklin) [reply distributed/deploy/adaptive.py]: Minimum here corresponds to the minimum number of workers that we want to keep.  The `workers_to_close` function is a slightly nicer place to handle this than outside.  For example if `workers_to_close` tells us that we need to close all of our workers, we then look at minimum and see that we need one, which worker should we keep?  Also, which *other* workers should we keep that are in the same group.  Suddenly we have to replicate the grouping logic.

Maximum doesn't make sense because we're only reducing the number of workers, not increasing them.

File: distributed/deploy/local.py (hunk 2 above)
Comment (mrocklin) [reply distributed/deploy/local.py]: Something was slightly nicer if we switched to using self.sync here (this is a common pattern in the client).  I decided to clean this up as part of this PR.

**Comment by mrocklin** [comment]
I'm fine to merge this any time.  I'll give you a chance to respond to recent commentary but I'll merge after you give the go-ahead.

**Comment by jhamman** [comment]
Thanks @mrocklin - go ahead and merge whenever now.
